File: ios/FitFusion/Services/HealthReminders.swift
import Foundation
import UserNotifications

/// Schedules periodic health reminders (hydration, exercise, sleep) using
/// `UNUserNotificationCenter`. All requests use identifiers prefixed with
/// `health_reminder_` so they can be cancelled individually or as a group.
@MainActor
public final class HealthReminders {

    // MARK: - Hydration

    /// Schedules repeating hydration reminders every 2 hours.
    public static func scheduleHydrationReminder() async {
        let center = UNUserNotificationCenter.current()

        let content = UNMutableNotificationContent()
        content.title = "Time to hydrate! 💧"
        content.body = "Stay on track — drink a glass of water."
        content.sound = .default

        let trigger = UNTimeIntervalNotificationTrigger(
            timeInterval: 2 * 60 * 60, // 2 hours
            repeats: true
        )

        let request = UNNotificationRequest(
            identifier: "health_reminder_hydration",
            content: content,
            trigger: trigger
        )
        try? await center.add(request)
    }

    // MARK: - Exercise

    /// Schedules a daily exercise reminder at the given hour (default 7 AM).
    public static func scheduleExerciseReminder(hour: Int = 7, minute: Int = 0) async {
        let center = UNUserNotificationCenter.current()

        let content = UNMutableNotificationContent()
        content.title = "Ready for today's workout? 🏋️"
        content.body = "Your daily exercise session is waiting."
        content.sound = .default

        var components = DateComponents()
        components.hour = hour
        components.minute = minute

        let trigger = UNCalendarNotificationTrigger(
            dateMatching: components,
            repeats: true
        )

        let request = UNNotificationRequest(
            identifier: "health_reminder_exercise",
            content: content,
            trigger: trigger
        )
        try? await center.add(request)
    }

    // MARK: - Sleep

    /// Schedules a nightly wind-down reminder 30 minutes before bedtime.
    public static func scheduleSleepReminder(bedtimeHour: Int = 22, bedtimeMinute: Int = 0) async {
        let center = UNUserNotificationCenter.current()

        var bedtime = DateComponents()
        bedtime.hour = bedtimeHour
        bedtime.minute = bedtimeMinute

        let calendar = Calendar.current
        let refDate = calendar.date(from: bedtime) ?? Date()
        let reminderDate = calendar.date(byAdding: .minute, value: -30, to: refDate) ?? refDate
        let reminderComponents = calendar.dateComponents([.hour, .minute], from: reminderDate)

        let content = UNMutableNotificationContent()
        content.title = "Wind down time 🌙"
        content.body = "Your sleep goal is in 30 minutes."
        content.sound = .default

        let trigger = UNCalendarNotificationTrigger(
            dateMatching: reminderComponents,
            repeats: true
        )

        let request = UNNotificationRequest(
            identifier: "health_reminder_sleep",
            content: content,
            trigger: trigger
        )
        try? await center.add(request)
    }

    // MARK: - Cancellation

    public static func cancelHydration() {
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: ["health_reminder_hydration"])
    }

    public static func cancelExercise() {
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: ["health_reminder_exercise"])
    }

    public static func cancelSleep() {
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: ["health_reminder_sleep"])
    }

    public static func cancelAll() {
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: [
                "health_reminder_hydration",
                "health_reminder_exercise",
                "health_reminder_sleep",
            ])
    }
}

File: android/app/src/main/java/com/myhealth/app/notifications/HealthReminders.kt
package com.myhealth.app.notifications

import android.content.Context
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import java.util.Calendar
import java.util.concurrent.TimeUnit

/**
 * Schedules periodic health reminders via WorkManager.
 *
 * All work requests are tagged with [TAG] so they can be cancelled in bulk.
 * Each reminder type also gets its own unique work name to allow individual
 * cancellation without affecting the others.
 */
object HealthReminders {

    const val TAG = "health_reminder"

    private const val WORK_HYDRATION = "health_reminder_hydration"
    private const val WORK_EXERCISE = "health_reminder_exercise"
    private const val WORK_SLEEP = "health_reminder_sleep"

    /**
     * Hydration reminder — fires every 2 hours.
     * WorkManager handles Doze / standby for us.
     */
    fun scheduleHydrationReminder(context: Context) {
        val data = workDataOf(
            ReminderWorker.KEY_TYPE to "hydration",
            ReminderWorker.KEY_TITLE to "Time to hydrate! 💧",
            ReminderWorker.KEY_BODY to "Stay on track — drink a glass of water.",
        )
        val request = PeriodicWorkRequestBuilder<ReminderWorker>(2, TimeUnit.HOURS)
            .setInputData(data)
            .addTag(TAG)
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_HYDRATION,
            ExistingPeriodicWorkPolicy.KEEP,
            request,
        )
    }

    /**
     * Exercise reminder — fires once per day.
     * Default preferred workout time is 7 AM; the initial delay is computed
     * to align the first fire with that hour.
     */
    fun scheduleExerciseReminder(context: Context, preferredHour: Int = 7) {
        val initialDelay = delayUntilNextHour(preferredHour)
        val data = workDataOf(
            ReminderWorker.KEY_TYPE to "exercise",
            ReminderWorker.KEY_TITLE to "Ready for today's workout? 🏋️",
            ReminderWorker.KEY_BODY to "Your daily exercise session is waiting.",
        )
        val request = PeriodicWorkRequestBuilder<ReminderWorker>(1, TimeUnit.DAYS)
            .setInitialDelay(initialDelay, TimeUnit.MILLISECONDS)
            .setInputData(data)
            .addTag(TAG)
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_EXERCISE,
            ExistingPeriodicWorkPolicy.KEEP,
            request,
        )
    }

    /**
     * Sleep reminder — fires once per day, 30 minutes before the user's
     * target bedtime. Defaults to 10 PM (22:00) bedtime → fires at 9:30 PM.
     */
    fun scheduleSleepReminder(context: Context, bedtimeHour: Int = 22, bedtimeMinute: Int = 0) {
        val cal = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, bedtimeHour)
            set(Calendar.MINUTE, bedtimeMinute)
            add(Calendar.MINUTE, -30)
        }
        val reminderHour = cal.get(Calendar.HOUR_OF_DAY)
        val reminderMinute = cal.get(Calendar.MINUTE)

        val initialDelay = delayUntilNextTime(reminderHour, reminderMinute)
        val data = workDataOf(
            ReminderWorker.KEY_TYPE to "sleep",
            ReminderWorker.KEY_TITLE to "Wind down time 🌙",
            ReminderWorker.KEY_BODY to "Your sleep goal is in 30 minutes.",
        )
        val request = PeriodicWorkRequestBuilder<ReminderWorker>(1, TimeUnit.DAYS)
            .setInitialDelay(initialDelay, TimeUnit.MILLISECONDS)
            .setInputData(data)
            .addTag(TAG)
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_SLEEP,
            ExistingPeriodicWorkPolicy.KEEP,
            request,
        )
    }

    fun cancelHydration(context: Context) {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_HYDRATION)
    }

    fun cancelExercise(context: Context) {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_EXERCISE)
    }

    fun cancelSleep(context: Context) {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_SLEEP)
    }

    fun cancelAll(context: Context) {
        WorkManager.getInstance(context).cancelAllWorkByTag(TAG)
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    private fun delayUntilNextHour(hour: Int): Long = delayUntilNextTime(hour, 0)

    private fun delayUntilNextTime(hour: Int, minute: Int): Long {
        val now = Calendar.getInstance()
        val target = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, hour)
            set(Calendar.MINUTE, minute)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (target.before(now)) target.add(Calendar.DAY_OF_YEAR, 1)
        return target.timeInMillis - now.timeInMillis
    }
}
